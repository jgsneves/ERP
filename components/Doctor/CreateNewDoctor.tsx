import {
  Button,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Input,
  Select,
  useToast,
  VStack,
} from "@chakra-ui/react";
import { EstadoCivil, Pessoas } from "@prisma/client";
import React, { useState } from "react";
import { v4 as uuid4 } from "uuid";
import { formatCPF } from "@brazilian-utils/brazilian-utils";
import axios from "axios";
import { useRouter } from "next/router";
import { DateFormat } from "utils/DateFormat";
import { ErrorHandler } from "utils/ErrorHandler";
import { CellphoneNumberFormat } from "utils/CellphoneNumberFormat";
import useCreateDoctorFormValidator from "hooks/useCreateDoctorFormValidator";

export default function CreateNewDoctor() {
  const [formData, setFormData] = useState<Pessoas>({
    Cpf: "",
    Email: "",
    EstadoCivil: "SOLTEIRO",
    Nacionalidade: "",
    PIS: "",
    RG: "",
    TelefonePessoal: "",
    TituloEleitor: "",
    CriadoEm: new Date(),
    Crm: "",
    DataNascimento: new Date(),
    EmpresaMedicaId: null,
    EnderecoId: null,
    Id: uuid4(),
    ModificadoEm: null,
    Nome: "",
    Participacao: null,
    Salario: null,
    StatusAdmissao: null,
    Tipo: "MEDICO",
    ModalidadeTrabalho: null,
  });
  const [dateInputValue, setDateInputValue] = useState<string>("");
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const toast = useToast();
  const router = useRouter();
  const {
    isCpfValid,
    isEmailValid,
    isFormValid,
    isNomeValid,
    verifyCpf,
    verifyEmail,
    verifyName,
    verifyForm,
  } = useCreateDoctorFormValidator(formData);

  const handleInputOnChange = (
    event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    const { id, value } = event.target;
    setFormData((state) => ({ ...state, [id]: value }));
  };

  const handleDateOnBlur = () => {
    if (dateInputValue) {
      const date = DateFormat.getDateTypeFromChakraString(dateInputValue);
      setFormData((state) => ({ ...state, DataNascimento: date }));
    }
  };

  const handleSubmitOnClick = () => {
    verifyForm();

    if (isFormValid) {
      setIsLoading(true);

      axios
        .post("/api/medicos", formData)
        .then(
          () => {
            toast({
              title: "Médico criado com sucesso!",
              duration: 5000,
              status: "success",
            });
            router.push("/medicos");
          },
          () => {
            toast({
              title: "Não foi possível criar um novo médico!",
              duration: 9000,
              isClosable: true,
              status: "error",
            });
          }
        )
        .catch((error) => ErrorHandler.logAxiosPostError(error))
        .finally(() => setIsLoading(false));
    }
  };

  return (
    <VStack alignItems="flex-start" maxW="500px">
      <FormControl isInvalid={!isNomeValid} isRequired>
        <FormLabel>Nome</FormLabel>
        <Input
          isDisabled={isLoading}
          value={formData.Nome}
          id="Nome"
          onChange={handleInputOnChange}
          onBlur={() => verifyName()}
        />
        {!isNomeValid && (
          <FormErrorMessage>Nome é um campo obrigatório.</FormErrorMessage>
        )}
      </FormControl>

      <FormControl isInvalid={!isEmailValid} isRequired>
        <FormLabel>E-mail:</FormLabel>
        <Input
          value={formData.Email}
          id="Email"
          onChange={handleInputOnChange}
          type="email"
          onBlur={() => verifyEmail()}
        />
        {!isEmailValid && (
          <FormErrorMessage>Insira um e-mail válido.</FormErrorMessage>
        )}
      </FormControl>

      <FormControl isInvalid={!isCpfValid} isRequired>
        <FormLabel>CPF</FormLabel>
        <Input
          isDisabled={isLoading}
          value={formatCPF(formData.Cpf)}
          id="Cpf"
          onChange={handleInputOnChange}
          onBlur={() => verifyCpf()}
        />
        {!isCpfValid && (
          <FormErrorMessage>Insira um CPF válido.</FormErrorMessage>
        )}
      </FormControl>
      <FormLabel>
        Telefone pessoal:
        <Input
          value={CellphoneNumberFormat.format(formData.TelefonePessoal)}
          id="TelefonePessoal"
          onChange={handleInputOnChange}
        />
      </FormLabel>
      <FormLabel>
        CRM
        <Input
          isDisabled={isLoading}
          value={formData.Crm ?? ""}
          id="Crm"
          onChange={handleInputOnChange}
        />
      </FormLabel>
      <FormLabel>
        RG:
        <Input value={formData.RG} id="RG" onChange={handleInputOnChange} />
      </FormLabel>

      <FormLabel>
        PIS:
        <Input
          value={formData.PIS ?? ""}
          id="PIS"
          onChange={handleInputOnChange}
        />
      </FormLabel>
      <FormLabel>
        Date de nascimento
        <Input
          isDisabled={isLoading}
          type="date"
          value={dateInputValue}
          onChange={(event) => setDateInputValue(event.target.value)}
          onBlur={handleDateOnBlur}
        />
      </FormLabel>
      <FormLabel>
        Nacionalidade:
        <Input
          value={formData.Nacionalidade}
          id="Nacionalidade"
          onChange={handleInputOnChange}
        />
      </FormLabel>

      <FormLabel>
        Estado civil:
        <Select
          value={formData.EstadoCivil}
          onChange={handleInputOnChange}
          id="EstadoCivil"
        >
          {Object.values(EstadoCivil).map((estadoCivil) => (
            <option key={estadoCivil} value={estadoCivil}>
              {estadoCivil}
            </option>
          ))}
        </Select>
      </FormLabel>
      <Button
        isLoading={isLoading}
        colorScheme="green"
        mt={2}
        onClick={handleSubmitOnClick}
      >
        salvar
      </Button>
    </VStack>
  );
}
