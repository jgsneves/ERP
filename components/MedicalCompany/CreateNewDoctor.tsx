import {
  Button,
  FormControl,
  FormLabel,
  Input,
  useToast,
  VStack,
} from "@chakra-ui/react";
import { Pessoas } from "@prisma/client";
import React, { useState } from "react";
import { v4 as uuid4 } from "uuid";
import { formatCPF } from "@brazilian-utils/brazilian-utils";
import { DateFormat } from "../../utils/dateFormat";
import axios from "axios";
import { useRouter } from "next/router";

interface Props {
  empresaId: string;
}

export default function CreateNewDoctor({ empresaId }: Props) {
  const [formData, setFormData] = useState<Pessoas>({
    ContaCorrenteId: null,
    Cpf: "",
    CriadoEm: new Date(),
    Crm: "",
    DataNascimento: new Date(),
    EmpresaMedicaId: empresaId,
    EnderecoId: null,
    Id: uuid4(),
    ModificadoEm: null,
    Nome: "",
    Participacao: null,
    Salario: null,
    StatusAdmissao: null,
    Tipo: "MEDICO",
  });
  const [dateInputValue, setDateInputValue] = useState<string>("");
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const toast = useToast();
  const router = useRouter();

  const handleInputOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { id, value } = event.target;
    setFormData((state) => ({ ...state, [id]: value }));
  };

  const handleDateOnBlur = () => {
    if (dateInputValue) {
      const date = DateFormat.getDateTypeFromChakraString(dateInputValue);
      setFormData((state) => ({ ...state, DataNascimento: date }));
    }
  };

  const handleSubmitOnClick = async () => {
    setIsLoading(true);

    await axios
      .post("/api/medicos", formData)
      .then(() => {
        toast({
          title: "Médico criado com sucesso!",
          duration: 5000,
          status: "success",
        });
        router.reload();
      })
      .catch((error) =>
        toast({
          title: "Não foi possível criar um novo médico!",
          description: error,
          duration: 9000,
          isClosable: true,
          status: "error",
        })
      )
      .finally(() => setIsLoading(false));
  };

  return (
    <VStack alignItems="flex-start">
      <FormControl maxW="500px">
        <FormLabel>
          Nome
          <Input
            isDisabled={isLoading}
            value={formData.Nome}
            id="Nome"
            onChange={handleInputOnChange}
          />
        </FormLabel>
        <FormLabel>
          CPF
          <Input
            isDisabled={isLoading}
            value={formatCPF(formData.Cpf)}
            id="Cpf"
            onChange={handleInputOnChange}
          />
        </FormLabel>
        <FormLabel>
          CRM
          <Input
            isDisabled={isLoading}
            value={formData.Crm ?? ""}
            id="Crm"
            onChange={handleInputOnChange}
          />
        </FormLabel>
        <FormLabel>
          Date de nascimento
          <Input
            isDisabled={isLoading}
            type="date"
            value={dateInputValue}
            onChange={(event) => setDateInputValue(event.target.value)}
            onBlur={handleDateOnBlur}
          />
        </FormLabel>
        <Button
          isLoading={isLoading}
          colorScheme="green"
          mt={2}
          onClick={handleSubmitOnClick}
        >
          salvar
        </Button>
      </FormControl>
    </VStack>
  );
}
