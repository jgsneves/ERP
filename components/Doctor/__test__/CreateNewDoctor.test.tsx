import { render } from "@testing-library/react";
import CreateNewDoctor from "../CreateNewDoctor";

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

describe("CreateNewDoctor", () => {
  it("should rende proper snapshot", () => {
    const { asFragment } = render(<CreateNewDoctor />);
    expect(asFragment()).toMatchSnapshot();
  });
});
