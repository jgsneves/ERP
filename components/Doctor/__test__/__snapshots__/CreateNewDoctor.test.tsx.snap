// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewDoctor should rende proper snapshot 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1behfr5"
  >
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="field-:r0:"
        id="field-:r0:-label"
      >
        Nome
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-0"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required="true"
        class="chakra-input css-0"
        id="Nome"
        required=""
        value=""
      />
    </div>
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="field-:r1:"
        id="field-:r1:-label"
      >
        E-mail:
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-0"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required="true"
        class="chakra-input css-0"
        id="Email"
        required=""
        type="email"
        value=""
      />
    </div>
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="field-:r2:"
        id="field-:r2:-label"
      >
        CPF
        <span
          aria-hidden="true"
          class="chakra-form__required-indicator css-0"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required="true"
        class="chakra-input css-0"
        id="Cpf"
        required=""
        value=""
      />
    </div>
    <label
      class="chakra-form__label css-uvho5g"
    >
      Telefone pessoal:
      <input
        class="chakra-input css-0"
        id="TelefonePessoal"
        value=""
      />
    </label>
    <label
      class="chakra-form__label css-uvho5g"
    >
      CRM
      <input
        class="chakra-input css-0"
        id="Crm"
        value=""
      />
    </label>
    <label
      class="chakra-form__label css-uvho5g"
    >
      RG:
      <input
        class="chakra-input css-0"
        id="RG"
        value=""
      />
    </label>
    <label
      class="chakra-form__label css-uvho5g"
    >
      PIS:
      <input
        class="chakra-input css-0"
        id="PIS"
        value=""
      />
    </label>
    <label
      class="chakra-form__label css-uvho5g"
    >
      Date de nascimento
      <input
        class="chakra-input css-0"
        type="date"
        value=""
      />
    </label>
    <label
      class="chakra-form__label css-uvho5g"
    >
      Nacionalidade:
      <input
        class="chakra-input css-0"
        id="Nacionalidade"
        value=""
      />
    </label>
    <label
      class="chakra-form__label css-uvho5g"
    >
      Estado civil:
      <div
        class="chakra-select__wrapper css-42b2qy"
      >
        <select
          class="chakra-select css-1ik61og"
          id="EstadoCivil"
        >
          <option
            value="SOLTEIRO"
          >
            SOLTEIRO
          </option>
          <option
            value="CASADO"
          >
            CASADO
          </option>
          <option
            value="SEPARADO"
          >
            SEPARADO
          </option>
          <option
            value="DIVORCIADO"
          >
            DIVORCIADO
          </option>
          <option
            value="VIUVO"
          >
            VIUVO
          </option>
        </select>
        <div
          class="chakra-select__icon-wrapper css-162mkon"
        >
          <svg
            aria-hidden="true"
            class="chakra-select__icon"
            focusable="false"
            role="presentation"
            style="width: 1em; height: 1em; color: currentColor;"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </label>
    <button
      class="chakra-button css-12eh3jv"
      type="button"
    >
      salvar
    </button>
  </div>
</DocumentFragment>
`;
